Introduce _WKSerializedNode

achristensen07 opened the pull request:
#### fba165ad3ce6fc4d987ae587ff526ce7ff0c8bea
<pre>
Introduce _WKSerializedNode
<a href="https://bugs.webkit.org/show_bug.cgi?id=296207">https://bugs.webkit.org/show_bug.cgi?id=296207</a>
<a href="https://rdar.apple.com/156173157">rdar://156173157</a>

Reviewed by Ryosuke Niwa.

This is an opaque type that can be used to clone nodes from one WKWebView to another.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::serializeNode const):
* Source/WebCore/dom/Attr.h:
* Source/WebCore/dom/CDATASection.cpp:
(WebCore::CDATASection::serializeNode const):
* Source/WebCore/dom/CDATASection.h:
* Source/WebCore/dom/ClonedNode.h: Added.
* Source/WebCore/dom/Comment.cpp:
(WebCore::Comment::serializeNode const):
* Source/WebCore/dom/Comment.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::serializeNode const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::serializeNode const):
* Source/WebCore/dom/DocumentFragment.h:
* Source/WebCore/dom/DocumentType.cpp:
(WebCore::DocumentType::serializeNode const):
* Source/WebCore/dom/DocumentType.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::serializeNode const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::deserializeNode):
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::serializeNode const):
* Source/WebCore/dom/ProcessingInstruction.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::serializeNode const):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/Text.cpp:
(WebCore::Text::serializeNode const):
* Source/WebCore/dom/Text.h:
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::serializeNode const):
* Source/WebCore/html/HTMLTemplateElement.h:
* Source/WebCore/page/WebKitClonedNode.cpp: Copied from Source/WebCore/page/WebKitNamespace.idl.
(WebCore::WebKitClonedNode::WebKitClonedNode):
* Source/WebCore/page/WebKitClonedNode.h: Copied from Source/WebCore/page/WebKitNamespace.idl.
(WebCore::WebKitClonedNode::create):
(WebCore::WebKitClonedNode::serializedNode const):
* Source/WebCore/page/WebKitClonedNode.idl: Copied from Source/WebCore/page/WebKitNamespace.idl.
* Source/WebCore/page/WebKitNamespace.cpp:
(WebCore::WebKitNamespace::createNodeInfo):
(WebCore::WebKitNamespace::cloneNode):
* Source/WebCore/page/WebKitNamespace.h:
* Source/WebCore/page/WebKitNamespace.idl:
* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::toAPI):
(WebKit::roundTripThroughSerializedScriptValue):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::JavaScriptEvaluationResult::toJS):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::toID):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::isSerializable):
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIClonedNode.cpp: Copied from Source/WebCore/page/WebKitNamespace.idl.
(API::ClonedNode::ClonedNode):
* Source/WebKit/UIProcess/API/APIClonedNode.h: Copied from Source/WebKit/UIProcess/Cocoa/WKColorExtensionView.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKClonedNode.h: Copied from Source/WebCore/page/WebKitNamespace.idl.
* Source/WebKit/UIProcess/API/Cocoa/_WKClonedNode.mm: Copied from Source/WebCore/page/WebKitNamespace.idl.
(-[_WKClonedNode dealloc]):
(-[_WKClonedNode _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKClonedNodeInternal.h: Copied from Source/WebCore/page/WebKitNamespace.idl.
* Source/WebKit/UIProcess/Cocoa/WKColorExtensionView.h:
* Source/WebKit/UIProcess/Cocoa/WKColorExtensionView.mm:
* Source/WebKit/UIProcess/mac/WKTextAnimationManagerMac.mm:
* Source/WebKit/UIProcess/mac/WebPopupMenuProxyMac.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm:
(TestWebKitAPI::TEST(ClonedNode, Basic)):

Canonical link: <a href="https://commits.webkit.org/297635@main">https://commits.webkit.org/297635@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1a758eee3199a16efc57281c0ecd565144bf894

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112388 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32120 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22597 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118467 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62771 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114350 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32772 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40683 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/85378 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115335 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/26157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/101115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65809 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/25452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/19253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62312 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/95541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/19328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121793 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39462 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29380 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/94187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39843 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/97355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/94012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39256 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/17053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18112 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39350 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/44838 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38985 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42322 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40728 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/ClonedNode.h:
```diff
@@ -0,0 +1,36 @@
+
+namespace WebCore {
+
+struct ClonedNode {
```
This isn't really clone but more like serialization format so I'd call this `SerializedNode` or `NodeSerialization` or something.

rniwa commented on Source/WebCore/page/WebKitClonedNode.idl:
```diff
@@ -0,0 +1,31 @@
+    EnabledForWorld=nodeInfoEnabled,
+    Exposed=Window,
+    ExportMacro=WEBCORE_EXPORT
+] interface WebKitClonedNode {
```
What is this object about? Is it supposed to represent serialized node??
Or so, it should be something like `WebKitSerializedNode`.

rniwa commented on Source/WebCore/page/WebKitNamespace.cpp:
```diff
@@ -66,11 +67,17 @@ UserMessageHandlersNamespace* WebKitNamespace::messageHandlers()
     return WebKitNodeInfo::create(node);
 }
 
+Ref<WebKitClonedNode> WebKitNamespace::cloneNode(Node& node, bool)
```
I wouldn't call this cloneNode. It's more like `serializeNode`.

rniwa commented on Source/WebCore/page/WebKitNamespace.idl:
```diff
@@ -32,4 +32,5 @@
 ] interface WebKitNamespace {
     readonly attribute UserMessageHandlersNamespace messageHandlers;
     [EnabledForWorld=nodeInfoEnabled] WebKitNodeInfo createNodeInfo(Node node);
+    [EnabledForWorld=nodeInfoEnabled] WebKitClonedNode cloneNode(Node node, optional boolean deep = false);
```
I don't think we should have a boolean argument like this.
Either we should add a new Options dictionary or just assume deep is true.

rniwa commented on Source/WebKit/UIProcess/API/APIClonedNode.cpp:
```diff
@@ -0,0 +1,38 @@
+
+namespace API {
+
+ClonedNode::ClonedNode(WebCore::SerializedNode&& serializedNode)
```
I'd call this SerializedNode instead for consistency.

rniwa commented on Source/WebKit/UIProcess/API/APIClonedNode.h:
```diff
@@ -0,0 +1,48 @@
+    static Ref<ClonedNode> create(WebCore::SerializedNode&& serializedNode) { return adoptRef(*new ClonedNode(WTFMove(serializedNode))); }
+    virtual ~ClonedNode();
+
+    const WebCore::SerializedNode& serializedNode() const { return m_serializedNode; }
```
maybe call this `coreSerializedNode` or something?

achristensen07 committed:
Introduce _WKSerializedNode
https://bugs.webkit.org/show_bug.cgi?id=296207
rdar://156173157

Reviewed by Ryosuke Niwa.

This is an opaque type that can be used to clone nodes from one WKWebView to another.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::serializeNode const):
* Source/WebCore/dom/Attr.h:
* Source/WebCore/dom/CDATASection.cpp:
(WebCore::CDATASection::serializeNode const):
* Source/WebCore/dom/CDATASection.h:
* Source/WebCore/dom/ClonedNode.h: Added.
* Source/WebCore/dom/Comment.cpp:
(WebCore::Comment::serializeNode const):
* Source/WebCore/dom/Comment.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::serializeNode const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::serializeNode const):
* Source/WebCore/dom/DocumentFragment.h:
* Source/WebCore/dom/DocumentType.cpp:
(WebCore::DocumentType::serializeNode const):
* Source/WebCore/dom/DocumentType.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::serializeNode const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::deserializeNode):
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::serializeNode const):
* Source/WebCore/dom/ProcessingInstruction.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::serializeNode const):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/Text.cpp:
(WebCore::Text::serializeNode const):
* Source/WebCore/dom/Text.h:
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::serializeNode const):
* Source/WebCore/html/HTMLTemplateElement.h:
* Source/WebCore/page/WebKitClonedNode.cpp: Copied from Source/WebCore/page/WebKitNamespace.idl.
(WebCore::WebKitClonedNode::WebKitClonedNode):
* Source/WebCore/page/WebKitClonedNode.h: Copied from Source/WebCore/page/WebKitNamespace.idl.
(WebCore::WebKitClonedNode::create):
(WebCore::WebKitClonedNode::serializedNode const):
* Source/WebCore/page/WebKitClonedNode.idl: Copied from Source/WebCore/page/WebKitNamespace.idl.
* Source/WebCore/page/WebKitNamespace.cpp:
(WebCore::WebKitNamespace::createNodeInfo):
(WebCore::WebKitNamespace::cloneNode):
* Source/WebCore/page/WebKitNamespace.h:
* Source/WebCore/page/WebKitNamespace.idl:
* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::toAPI):
(WebKit::roundTripThroughSerializedScriptValue):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::JavaScriptEvaluationResult::toJS):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::toID):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::isSerializable):
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIClonedNode.cpp: Copied from Source/WebCore/page/WebKitNamespace.idl.
(API::ClonedNode::ClonedNode):
* Source/WebKit/UIProcess/API/APIClonedNode.h: Copied from Source/WebKit/UIProcess/Cocoa/WKColorExtensionView.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKClonedNode.h: Copied from Source/WebCore/page/WebKitNamespace.idl.
* Source/WebKit/UIProcess/API/Cocoa/_WKClonedNode.mm: Copied from Source/WebCore/page/WebKitNamespace.idl.
(-[_WKClonedNode dealloc]):
(-[_WKClonedNode _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKClonedNodeInternal.h: Copied from Source/WebCore/page/WebKitNamespace.idl.
* Source/WebKit/UIProcess/Cocoa/WKColorExtensionView.h:
* Source/WebKit/UIProcess/Cocoa/WKColorExtensionView.mm:
* Source/WebKit/UIProcess/mac/WKTextAnimationManagerMac.mm:
* Source/WebKit/UIProcess/mac/WebPopupMenuProxyMac.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm:
(TestWebKitAPI::TEST(ClonedNode, Basic)):

Canonical link: https://commits.webkit.org/297635@main
